File: android/app/src/main/java/com/betterlise/app/domain/SimulatorGrouping.kt
package com.betterlise.app.domain

import com.betterlise.app.data.api.Grade
import kotlinx.serialization.Serializable
import java.util.UUID

/** A hypothetical grade added by the user. */
@Serializable
data class SimulatedGrade(
    val id: String = UUID.randomUUID().toString(),
    val name: String,
    val grade: Double,
    val coeff: Double,
    val classCode: String,
)

/** What the user changed in the simulator, persisted on the device. */
@Serializable
data class SimulatorData(
    val simulations: List<SimulatedGrade> = emptyList(),
    /** Coefficients typed by the user, by grade code. They win over community ones. */
    val localCoeffs: Map<String, Double> = emptyMap(),
    /** UE chosen by the user for grades whose code doesn't parse, by grade code. */
    val classOverrides: Map<String, String> = emptyMap(),
)

data class SimulatorRealGrade(
    val grade: Grade,
    /** Community coefficient, or 1 without votes. */
    val baseCoeff: Double,
    val isCommunity: Boolean,
    val effectiveCoeff: Double,
) {
    val code: String get() = grade.code
    /** The user's coefficient differs from the shared one: it can be shared. */
    val canShare: Boolean get() = effectiveCoeff != baseCoeff
}

data class UEGroup(
    val classCode: String,
    val semester: String,
    val real: List<SimulatorRealGrade>,
    val simulations: List<SimulatedGrade>,
) {
    private val realItems get() = real.map { it.grade.note to it.effectiveCoeff }
    val currentAverage: Double get() = SimulatorGrouping.mean(realItems)
    val projectedAverage: Double get() = SimulatorGrouping.mean(realItems + simulations.map { it.grade to it.coeff })
    val hasSimulations: Boolean get() = simulations.isNotEmpty()
}

/** Grade simulator rules, ported from the web (hooks/useGradeSimulation.ts). */
object SimulatorGrouping {
    const val ALL_SEMESTERS = "all"
    const val SIMULATED_SEMESTER = "Sim"

    /** Semesters found in the grade codes, most recent first. */
    fun availableSemesters(grades: List<Grade>): List<String> =
        grades.map { ClassCodeParser.parse(it.code).semester }
            .filter { it != ClassCodeParser.UNASSIGNED }
            .distinct()
            .sortedByDescending(ClassCodeParser::semesterNumber)

    fun groups(grades: List<Grade>, weights: Map<String, Double>, data: SimulatorData, semester: String): List<UEGroup> {
        val semesters = LinkedHashMap<String, String>()
        val real = mutableMapOf<String, MutableList<SimulatorRealGrade>>()
        val simulations = mutableMapOf<String, MutableList<SimulatedGrade>>()

        fun register(classCode: String, groupSemester: String, isReal: Boolean) {
            val current = semesters[classCode]
            if (current == null) {
                semesters[classCode] = groupSemester
            } else if (isReal && current == ClassCodeParser.UNASSIGNED && groupSemester != ClassCodeParser.UNASSIGNED) {
                semesters[classCode] = groupSemester
            }
        }

        for (grade in grades) {
            val parsed = ClassCodeParser.parse(grade.code)
            val included = semester == ALL_SEMESTERS || parsed.semester == semester || parsed.semester == ClassCodeParser.UNASSIGNED
            if (!included) continue
            val classCode = data.classOverrides[grade.code] ?: parsed.classCode
            register(classCode, parsed.semester, isReal = true)
            val community = weights[grade.code]
            val baseCoeff = community ?: 1.0
            real.getOrPut(classCode, ::mutableListOf) += SimulatorRealGrade(
                grade = grade,
                baseCoeff = baseCoeff,
                isCommunity = community != null,
                effectiveCoeff = data.localCoeffs[grade.code] ?: baseCoeff,
            )
        }

        for (simulation in data.simulations) {
            val classCode = simulation.classCode.ifEmpty { ClassCodeParser.UNASSIGNED }
            register(classCode, SIMULATED_SEMESTER, isReal = false)
            simulations.getOrPut(classCode, ::mutableListOf) += simulation
        }

        return semesters.keys.sorted().map {
            UEGroup(it, semesters.getValue(it), real[it].orEmpty(), simulations[it].orEmpty())
        }
    }

    /** Weighted mean, 0 when coefficients sum to 0. */
    fun mean(items: List<Pair<Double, Double>>): Double {
        val totalCoeff = items.sumOf { it.second }
        if (totalCoeff == 0.0) return 0.0
        return items.sumOf { it.first * it.second } / totalCoeff
    }
}

File: ios/BetterLise/Features/Simulator/SimulatorGrouping.swift
import Foundation

/// A hypothetical grade added by the user.
struct SimulatedGrade: Codable, Hashable, Identifiable, Sendable {
    var id = UUID().uuidString
    let name: String
    var grade: Double
    let coeff: Double
    let classCode: String
}

/// What the user changed in the simulator, persisted on the device.
struct SimulatorData: Codable, Equatable, Sendable {
    var simulations: [SimulatedGrade] = []
    /// Coefficients typed by the user, by grade code. They win over community ones.
    var localCoeffs: [String: Double] = [:]
    /// UE chosen by the user for grades whose code doesn't parse, by grade code.
    var classOverrides: [String: String] = [:]
}

struct SimulatorRealGrade: Identifiable, Equatable, Sendable {
    let grade: Grade
    /// Community coefficient, or 1 without votes.
    let baseCoeff: Double
    let isCommunity: Bool
    let effectiveCoeff: Double

    var id: String { grade.code }
    /// The user's coefficient differs from the shared one: it can be shared.
    var canShare: Bool { effectiveCoeff != baseCoeff }
}

struct UEGroup: Identifiable, Equatable, Sendable {
    let classCode: String
    let semester: String
    let real: [SimulatorRealGrade]
    let simulations: [SimulatedGrade]

    var id: String { classCode }

    var currentAverage: Double {
        SimulatorGrouping.mean(real.map { ($0.grade.note, $0.effectiveCoeff) })
    }

    var projectedAverage: Double {
        SimulatorGrouping.mean(real.map { ($0.grade.note, $0.effectiveCoeff) } + simulations.map { ($0.grade, $0.coeff) })
    }

    var hasSimulations: Bool { !simulations.isEmpty }
}

/// Grade simulator rules, ported from the web (hooks/useGradeSimulation.ts).
enum SimulatorGrouping {
    static let allSemesters = "all"
    static let simulatedSemester = "Sim"

    /// Semesters found in the grade codes, most recent first.
    static func availableSemesters(_ grades: [Grade]) -> [String] {
        let semesters = Set(grades.map { ClassCodeParser.parse($0.code).semester })
            .subtracting([ClassCodeParser.unassigned])
        return semesters.sorted { ClassCodeParser.semesterNumber($0) > ClassCodeParser.semesterNumber($1) }
    }

    static func groups(
        grades: [Grade],
        weights: [String: Double],
        data: SimulatorData,
        semester: String
    ) -> [UEGroup] {
        var semesters: [String: String] = [:]
        var real: [String: [SimulatorRealGrade]] = [:]
        var simulations: [String: [SimulatedGrade]] = [:]
        var order: [String] = []

        func register(_ classCode: String, semester: String, isReal: Bool) {
            if semesters[classCode] == nil {
                semesters[classCode] = semester
                order.append(classCode)
            } else if isReal, semesters[classCode] == ClassCodeParser.unassigned, semester != ClassCodeParser.unassigned {
                semesters[classCode] = semester
            }
        }

        for grade in grades {
            let parsed = ClassCodeParser.parse(grade.code)
            guard semester == allSemesters || parsed.semester == semester || parsed.semester == ClassCodeParser.unassigned else {
                continue
            }
            let classCode = data.classOverrides[grade.code] ?? parsed.classCode
            register(classCode, semester: parsed.semester, isReal: true)
            let community = weights[grade.code]
            let baseCoeff = community ?? 1
            real[classCode, default: []].append(SimulatorRealGrade(
                grade: grade,
                baseCoeff: baseCoeff,
                isCommunity: community != nil,
                effectiveCoeff: data.localCoeffs[grade.code] ?? baseCoeff
            ))
        }

        for simulation in data.simulations {
            let classCode = simulation.classCode.isEmpty ? ClassCodeParser.unassigned : simulation.classCode
            register(classCode, semester: simulatedSemester, isReal: false)
            simulations[classCode, default: []].append(simulation)
        }

        return order.sorted().map {
            UEGroup(classCode: $0, semester: semesters[$0] ?? ClassCodeParser.unassigned, real: real[$0] ?? [], simulations: simulations[$0] ?? [])
        }
    }

    /// Weighted mean, 0 when coefficients sum to 0.
    static func mean(_ items: [(grade: Double, coeff: Double)]) -> Double {
        let totalCoeff = items.reduce(0) { $0 + $1.coeff }
        guard totalCoeff != 0 else { return 0 }
        return items.reduce(0) { $0 + $1.grade * $1.coeff } / totalCoeff
    }
}
